aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py

```diff
@@ -0,0 +1,823 @@
+            )
+        except Exception as err:
+            raise PresentationExchError(
+                "Error creating requirement " f"inside to_requirement function, {err}"
+            )
+    return requirement
+
+
+async def is_len_applicable(req: Requirement, val: int) -> bool:
+    """
+    Check and validate requirement minimum, maximum and count.
+
+    Args:
+        req: Requirement
+        val: int value to check
+    Return:
+        bool
+
+    """
+    if req.count:
+        if req.count > 0 and val != req.count:
+            return False
+    if req.minimum:
+        if req.minimum > 0 and req.minimum > val:
+            return False
+    if req.maximum:
+        if req.maximum > 0 and req.maximum < val:
+            return False
+    return True
+
+
+async def contains(data: Sequence[str], e: str) -> bool:
+    """
+    Check for e in data.
+
+    Returns True if e exists in data else return False
+
+    Args:
+        data: Sequence of str
+        e: str value to check
+    Return:
+        bool
+
+    """
+    data_list = list(data) if data else []
+    for tmp_item in data_list:
+        if e == tmp_item:
+            return True
+    return False
+
+
+async def filter_constraints(
+    constraints: Constraints, credentials: Sequence[VCRecord]
+) -> Sequence[VCRecord]:
+    """
+    Return list of applicable VCRecords after applying filtering.
+
+    Args:
+        constraints: Constraints
+        credentials: Sequence of credentials
+            to apply filtering on
+    Return:
+        Sequence of applicable VCRecords
+
+    """
+    result = []
+    for tmp_cred in credentials:
+        if (
+            constraints.subject_issuer is not None
+            and constraints.subject_issuer == "required"
+            and not await subject_is_issuer(credential=tmp_cred)
+        ):
+            continue
+
+        applicable = False
+        predicate = False
+        for tmp_field in constraints._fields:
+            applicable = await filter_by_field(tmp_field, tmp_cred)
+            if tmp_field.predicate:
+                if tmp_field.predicate == "required":
+                    predicate = True
+            if applicable:
+                break
+        if not applicable:
+            continue
+
+        # TODO: create new credential with selective disclosure
+        if constraints.limit_disclosure or predicate:
+            raise PresentationExchError("Not yet implemented - createNewCredential")
+
+        result.append(tmp_cred)
+    return result
+
+
+async def filter_by_field(field: Field, credential: VCRecord) -> bool:
+    """
+    Apply filter on VCRecord.
+
+    Checks if a credential is applicable
+
+    Args:
+        field: Field contains filtering spec
+        credential: credential to apply filtering on
+    Return:
+        bool
+
+    """
+    for tmp_path in field.paths:
+        tmp_jsonpath = parse(tmp_path)
+        match = tmp_jsonpath.find(json.loads(credential.value))
```

Comment:
You're parsing the stringified credential for every iteration of the loop. Also from what I can see, `filter_by_field` is called multiple times for every credential.

Preferably the already parsed credential is passed. But we can at least call `json.loads` outside of the `fields.path` loop